Expo ImagePicker fails to launch. Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.

### Summary

My users on Android arrive at a point where they can no longer launch the camera.
User has given camera and media library permissions.
User was able to open camera.
User is no longer able to open Camera or Media Library.

It seems to be related to the expo task manager too.

** I've found that the issue appears when: ** 
- Given you are tracking location via a task in the task manager
- When you close the app
- When you reopen the app
- When you try to pick an image
- Then the image picker will not open

I've found that the issue does **not** appear when location tracking is not active when the app was last closed. 

Check the linked repo with reproducible example.


``` 
result = await ImagePicker.launchCameraAsync({
        base64: true,
        allowsEditing: true,
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        quality: 0.2
      });

// or

result = await ImagePicker.launchImageLibraryAsync({...})
```

Resulting error logs (source Sentry):
```
Error: Error: Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
→ Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.CameraContra...
  at ? (app:///index.android.bundle:3411:1320)
  at generatorResume([native code])
  at n(app:///index.android.bundle:269:69)
  at p(app:///index.android.bundle:269:316)
  at u(app:///index.android.bundle:2634:157)
  at ? (app:///index.android.bundle:2634:884)
  at ? (app:///index.android.bundle:76:1692)
  at p(app:///index.android.bundle:76:528)
  at N(app:///index.android.bundle:76:918)
  at callReactNativeMicrotasks(app:///index.android.bundle:76:3079)
  at value(app:///index.android.bundle:56:2935)
  at ? (app:///index.android.bundle:56:1045)
  at value(app:///index.android.bundle:56:2583)
  at value(app:///index.android.bundle:56:1004)
  at value([native code])
  at value([native code])
```

Closing and reopening the app does not resolve the issue.
Once a user has met this error, only reinstalling the app works.

### What platform(s) does this occur on?

Android

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.0.1
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 16.12.0 - ~/.nvm/versions/node/v16.12.0/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 8.1.0 - ~/.nvm/versions/node/v16.12.0/bin/npm
      Watchman: 2021.06.07.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.7199119
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^46.0.0 => 46.0.9 
      react: 18.0.0 => 18.0.0 
      react-dom: 18.0.0 => 18.0.0 
      react-native: 0.69.6 => 0.69.6 
      react-native-web: ~0.18.7 => 0.18.8 
      react-navigation: ^4.4.4 => 4.4.4 
    npmGlobalPackages:
      eas-cli: 0.53.1
      expo-cli: 6.0.6
    Expo Workflow: managed

### Minimal reproducible example

The error appears only for production builds and dev builds using a dev-client.  
I've been able to reproduce it via the example demo app in this github repository: https://github.com/Nxtra/expo-image-picker-issue-reproduce  

The steps to reproduce the exact issue are in the repo, also here:
- run `npx expo start`
- open via dev client
- Press "pick an image"
- You will be able to pick an image
- Select start tracking
- After a few sec an icon will appear in you top notification bar indicating that you are being tracked
- Close the app (really close it by swiping it out of your open apps menu)
- Open app
- Press "pick an image"
- **The camera picker won't open!**

I am running into this as well, however the error is on the image picker, not the camera:

`Call to function 'ExponentImagePicker.launchImageLibraryAsync' has been rejected.
→ Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.ImageLibraryContract@7b3ffcc and input ImageLibraryContractOptions(options=expo.modules.imagepicker.ImagePickerOptions@fbf9cf1). You must ensure the ActivityResultLauncher is registered before calling launch()`

Unlike the OP, I do not engage in background location tracking. I do make calls to fetch location and they might be underway while the user is attempting to pick a photo. Perhaps that's enough to trigger this?

Can confirm I get the same error when following the steps in the repro:
```
[Error: Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
→ Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.CameraContract@396e31d and input CameraContractOptions(uri=content://be.aleqsio.camera.ImagePickerFileProvider/cached_expo_files/ImagePicker/562d7df1-ae7a-40df-a5c8-68ca0857fe0c.jpeg, options=expo.modules.imagepicker.ImagePickerOptions@5dfb992). You must ensure the ActivityResultLauncher is registered before calling launch()]
```

@mauricedoepke Yeah both files `expo-image-picker+14.3.0.patch` and `expo-modules-core+1.5.3.patch`

I'm getting this error:
```
WARN  Possible Unhandled Promise Rejection (id: 0):
Error: Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
→ Caused by: java.lang.NullPointerException: Parameter specified as non-null is null: method host.exp.exponent.experience.d.checkSelfPermission, parameter permission
Error: Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
→ Caused by: java.lang.NullPointerException: Parameter specified as non-null is null: method host.exp.exponent.experience.d.checkSelfPermission, parameter permission
    at construct (native)
    at apply (native)
    at _construct (http://192.168.1.174:19000/node_modules%5Cexpo%5CAppEntry.bundle//&platform=android&dev=true&hot=false:3348:28)  
    at Wrapper (http://192.168.1.174:19000/node_modules%5Cexpo%5CAppEntry.bundle//&platform=android&dev=true&hot=false:3310:25)     
    at construct (native)
    at _createSuperInternal (http://192.168.1.174:19000/node_modules%5Cexpo%5CAppEntry.bundle//&platform=android&dev=true&hot=false:96614:322)
    at call (native)
    at CodedError (http://192.168.1.174:19000/node_modules%5Cexpo%5CAppEntry.bundle//&platform=android&dev=true&hot=false:96622:26)
```

I can confirm it's happening on SDK 49

"expo-image-picker": "~14.3.2",

```
Error: Call to function 'ExponentImagePicker.launchImageLibraryAsync' has been rejected.
→ Caused by: java.lang.IllegalArgumentException: Uri lacks 'file' scheme: content://media/picker/0/com.android.providers.media.photopicker/media/1000008735
```

When I add these patches and try to do an eas build, I get the following error:

Expo 49.

```
> Task :app:createBundleReleaseJsAndAssets_SentryUploadCleanUp SKIPPED
w: Detected multiple Kotlin daemon sessions at build/kotlin/sessions
999 actionable tasks: 999 executed
[stderr] FAILURE: Build failed with an exception.
[stderr] * What went wrong:
[stderr] Execution failed for task ':expo-image-picker:compileReleaseKotlin'.
[stderr] > A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
[stderr] > Compilation error. See log for more details
[stderr] * Try:
[stderr] > Run with --stacktrace option to get the stack trace.
[stderr] > Run with --info or --debug option to get more log output.
[stderr] > Run with --scan to get full insights.
[stderr] * Get more help at https://help.gradle.org
[stderr] BUILD FAILED in 9m 8s
Error: Gradle build failed with unknown error. See logs for the "Run gradlew" phase for more information.
```

This DOES NOT fix the issue

```
Call to function 'ExponentImagePicker.launchImageLibraryAsync' has been rejected.
→ Caused by: java.lang.IllegalArgumentException: Uri lacks 'file' scheme: content://media/picker/0/com.android.providers.media.photopicker/media/1000043213
```

Still there after applying the patch

Hi! I’m running into this on Expo 49.0.13 with both `launchCameraAsync` and `launchImageLibraryAsync`—seeing currently on Android 12 devices.

```
Error: Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
→ Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.CameraContract@af26...
  at → Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.CameraContract@af267e0 and input CameraContractOptions(uri=(app:///index.android.bundle)
```

```
Error: Call to function 'ExponentImagePicker.launchImageLibraryAsync' has been rejected.
→ Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.ImageLibraryC...
```

Is there any consensus on whether the patches work or not and what they might be missing given how long ago they were suggested and how long this bug has been open?

--- a/packages/expo-modules-core/CHANGELOG.md
+++ b/packages/expo-modules-core/CHANGELOG.md
@@ -16,7 +16,7 @@
 - Fixed breaking changes from React-Native 0.74. ([#26357](https://github.com/expo/expo/pull/26357) by [@kudo](https://github.com/kudo))
 - Fix proguard rules so `Serializable` types are not obfuscated. ([#26545](https://github.com/expo/expo/pull/26545) by [@alanjhughes](https://github.com/alanjhughes))
 - Fixed breaking changes from React-Native 0.74. ([#26357](https://github.com/expo/expo/pull/26357), [#26587](https://github.com/expo/expo/pull/26587) by [@kudo](https://github.com/kudo))
-- [Android] Fixed activity contract registration after host destruction.
+- [Android] Fixed activity contract registration after host destruction. ([#26881](https://github.com/expo/expo/pull/26881) by [@lukmccall](https://github.com/lukmccall))
 
 ### 💡 Others
 

--- a/packages/expo-modules-core/CHANGELOG.md
+++ b/packages/expo-modules-core/CHANGELOG.md
@@ -16,6 +16,7 @@
 - Fixed breaking changes from React-Native 0.74. ([#26357](https://github.com/expo/expo/pull/26357) by [@kudo](https://github.com/kudo))
 - Fix proguard rules so `Serializable` types are not obfuscated. ([#26545](https://github.com/expo/expo/pull/26545) by [@alanjhughes](https://github.com/alanjhughes))
 - Fixed breaking changes from React-Native 0.74. ([#26357](https://github.com/expo/expo/pull/26357), [#26587](https://github.com/expo/expo/pull/26587) by [@kudo](https://github.com/kudo))
+- [Android] Fixed activity contract registration after host destruction.
 
 ### 💡 Others
 

--- a/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt
+++ b/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt
@@ -63,6 +63,8 @@ class AppContext(
   val registry = ModuleRegistry(WeakReference(this))
   private val reactLifecycleDelegate = ReactLifecycleDelegate(this)
 
+  private var hostWereDestroyed = false
+
   // We postpone creating the `JSIInteropModuleRegistry` to not load so files in unit tests.
   internal lateinit var jsiInterop: JSIInteropModuleRegistry
 
@@ -138,6 +140,7 @@ class AppContext(
   fun onCreate() = trace("AppContext.onCreate") {
     registry.readyForPostingEvents()
     registry.post(EventName.MODULE_CREATE)
+    registry.registerActivityContracts()
     registry.flushTheEventQueue()
   }
 
@@ -329,6 +332,12 @@ class AppContext(
       "Current Activity is of incorrect class, expected AppCompatActivity, received ${currentActivity?.localClassName}"
     }
 
+    // We need to re-register activity contracts when reusing AppContext with new Activity after host destruction.
+    if (hostWereDestroyed) {
+      hostWereDestroyed = false
+      registry.registerActivityContracts()
+    }
+
     activityResultsManager.onHostResume(activity)
     registry.post(EventName.ACTIVITY_ENTERS_FOREGROUND)
   }
@@ -346,6 +355,9 @@ class AppContext(
       activityResultsManager.onHostDestroy(it)
     }
     registry.post(EventName.ACTIVITY_DESTROYS)
+    // The host (Activity) was destroyed, but it doesn't mean that modules will be destroyed too.
+    // So we save that information, and we will re-register activity contracts when the host will be resumed with new Activity.
+    hostWereDestroyed = true
   }
 
   internal fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {

--- a/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt
+++ b/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt
@@ -63,7 +63,7 @@ class AppContext(
   val registry = ModuleRegistry(WeakReference(this))
   private val reactLifecycleDelegate = ReactLifecycleDelegate(this)
 
-  private var hostWereDestroyed = false
+  private var hostWasDestroyed = false
 
   // We postpone creating the `JSIInteropModuleRegistry` to not load so files in unit tests.
   internal lateinit var jsiInterop: JSIInteropModuleRegistry
@@ -333,8 +333,8 @@ class AppContext(
     }
 
     // We need to re-register activity contracts when reusing AppContext with new Activity after host destruction.
-    if (hostWereDestroyed) {
-      hostWereDestroyed = false
+    if (hostWasDestroyed) {
+      hostWasDestroyed = false
       registry.registerActivityContracts()
     }
 
@@ -357,7 +357,7 @@ class AppContext(
     registry.post(EventName.ACTIVITY_DESTROYS)
     // The host (Activity) was destroyed, but it doesn't mean that modules will be destroyed too.
     // So we save that information, and we will re-register activity contracts when the host will be resumed with new Activity.
-    hostWereDestroyed = true
+    hostWasDestroyed = true
   }
 
   internal fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {

--- a/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ModuleRegistry.kt
+++ b/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ModuleRegistry.kt
@@ -33,10 +33,6 @@ class ModuleRegistry(
       )
     }
 
-    holder.apply {
-      registerContracts()
-    }
-
     registry[holder.name] = holder
   }
 
@@ -110,6 +106,12 @@ class ModuleRegistry(
   fun cleanUp() {
     registry.clear()
     logger.info("✅ ModuleRegistry was destroyed")
+  }
+
+  internal fun registerActivityContracts() {
+    forEach { holder ->
+      holder.registerContracts()
+    }
   }
 
   /**

